server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java

```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                failureCount.incrementAndGet();
+                logger.error("failure while getting future", e);
+            }
+        }
+        executorService.shutdown();
+        executorService.awaitTermination(10, TimeUnit.SECONDS);
+
+        if (failureCount.get() > 0 || successCount.get() != expectedExecutions || requestCount.get() != successCount.get()) {
+            fail("\nExpected Bulks: " + expectedExecutions + "\n" +
+                "Requested Bulks: " + requestCount.get() + "\n" +
+                "Successful Bulks: " + successCount.get() + "\n" +
+                "Failed Bulks: " + failureCount.get() + "\n" +
+                "Max Documents: " + maxDocuments + "\n" +
+                "Max Batch Size: " + maxBatchSize + "\n" +
+                "Concurrent Clients: " + concurrentClients + "\n" +
+                "Concurrent Bulk Requests: " + concurrentBulkRequests + "\n"
+            );
+        }
+        bulkProcessor.close();
+        //count total docs after processor is closed since there may have been partial batches that are flushed on close.
+        assertEquals(docCount.get(), maxDocuments);
+    }
+
+    public void testConcurrentExecutionsWithFlush() throws Exception {
+        final int maxDocuments = 100_000;
+        final int concurrentClients = 2;
+        final int maxBatchSize = Integer.MAX_VALUE; //don't flush based on size
+        final int concurrentBulkRequests = randomIntBetween(0, 20);
+        final int simulateWorkTimeInMillis = 5;
+        BulkResponse bulkResponse = new BulkResponse(new BulkItemResponse[]{new BulkItemResponse()}, 0);
+        AtomicInteger failureCount = new AtomicInteger(0);
+        AtomicInteger successCount = new AtomicInteger(0);
+        AtomicInteger requestCount = new AtomicInteger(0);
+        AtomicInteger docCount = new AtomicInteger(0);
+        BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer = (request, listener) ->
+        {
+            try {
+                Thread.sleep(simulateWorkTimeInMillis); //simulate work
+                listener.onResponse(bulkResponse);
+            } catch (InterruptedException e) {
+                failureCount.getAndIncrement();
+                logger.error("interrupted while sleeping. There is likely something wrong with this test!", e);
+            }
+        };
+        ScheduledExecutorService flushExecutor = Executors.newScheduledThreadPool(1);
+        BulkProcessor bulkProcessor = new BulkProcessor(consumer, BackoffPolicy.noBackoff(),
+            countingListener(requestCount, successCount, failureCount, docCount),
+            concurrentBulkRequests, maxBatchSize, new ByteSizeValue(Integer.MAX_VALUE),
+            TimeValue.timeValueMillis(simulateWorkTimeInMillis * 2),
+            (command, delay, executor) ->
+                Scheduler.wrapAsScheduledCancellable(flushExecutor
+                    .schedule(command, delay.millis(), TimeUnit.MILLISECONDS)), flushExecutor::shutdownNow, BulkRequest::new);
```

Comment:
I wonder if we should wait for an ongoing flush to terminate

```suggestion
                    .schedule(command, delay.millis(), TimeUnit.MILLISECONDS)), () -> {
                         flushExecutor.shutdown();
                         try {
                             flushExecutor.awaitTermination(30L, TimeUnit.SECONDS);
                             if (flushExecutor.isTerminated() == false) {
                                 flushExecutor.shutdownNow();
                             }
                         } catch (InterruptedException e) {
                             Thread.currentThread().interrupt();
                         }
                    }, BulkRequest::new);
```